app/advancedfilters/taxonFilter.js

```diff
@@ -90,25 +29,52 @@ app.controller('taxonFilter', function($scope, $q, hardCodedDataService,
 
   $scope.addTaxons = function() {
     var taxons = stringService.getTextareaItemsAsArray($scope.taxonTextArea);
-    var tempTaxa = {};
-    var validIds = [];
-    angular.forEach(taxons, function(taxonId){
-      if($scope.taxa[taxonId]) {
-        $scope.taxa[taxonId].checked = true;
-      } else {
-        if (validationService.validateTaxon(taxonId)) {
-          validIds.push(taxonId);
-          tempTaxa[taxonId] = {
-            'name': taxonId,
-            'checked': true
-          };
-        }
+    var items = filterService.addFilterItems(taxons, validationService.validateTaxon);
+    $scope.taxa = filterService.mergeRightToLeft(items, $scope.taxa);
+    updateTaxonInfo();
+    $scope.taxonTextArea = '';
+  };
+
+  $scope.removeTaxIds = function(idsToRemove, taxa) {
+    return _.filter(taxa, function(d){
+      return !_.contains(idsToRemove, parseInt(d.id));
+    });
+  }
+
+  var updateTaxonInfo = function() {
+    taxonomyService.getTaxa(_.pluck($scope.taxa,'id')).then(function(data){
+      filterService.enrichFilterItemObject($scope.taxa, data.data.taxonomies, 'taxonomyId');
+      if(data.data.errors) {
+        // remove from list
+        var obsoleteIds = _.pluck(data.data.errors, 'requestedId');
+        $rootScope.alerts = _.map(data.data.errors, function(message){
+          return {
+            msg: message.requestedId + ': ' + message.errorMessage
+          }
+        });
+        $scope.taxa = $scope.removeTaxIds(obsoleteIds, $scope.taxa);
+      }
+      if(data.data.redirects) {
+        // update object in list
+        $scope.taxa = redirectTaxa($scope.taxa, data.data.redirects);
+        updateTaxonInfo();
       }
     });
-    getTaxaInfo(validIds).then(function(data) {
-      cleanTaxaInformation(tempTaxa, data);
+  }
+
+
+  var redirectTaxa = function(taxaInfo, redirections) {
+    var redirectedIds = [];
+    var redirectionMap = _.indexBy(redirections, 'requestedId');
+    return _.map(taxaInfo, function(d){
+      if(redirectionMap[d.id]) {
+        var updatedId = redirectionMap[d.id].redirectLocation.substring(redirectionMap[d.id].redirectLocation.lastIndexOf('/')+1);
+        $rootScope.alerts.push({msg: d.id + ' was updated to ' + updatedId});
+        d.id = updatedId;
+      }
+      return d;
     });
-    $scope.taxonTextArea = '';
+    return redirectionMap;
```

Comment:
You have a return statement before this one, so this one is not reachable, please double check this function.